### ISSUE TITLE
proposal - reload config file on save

### ISSUE THREAD

**Issue by YanB25** [fatedier/frp#1176]
## abstract
I'd like to contribute to this project by adding a new feature. I wish to support reloading the config files automatically on save.
## details
Users can enable this feature in `frpc.ini`
```
reload_on_save = true
```
and whenever the config file is written, a `reload` is automatically triggered.
### implementation details
I would use [https://github.com/fsnotify/fsnotify](fsnotify)(Cross-platform file system notifications for Go) to watch the config files and use `go's channel` to pass the information asynchronously.
## motivation
It's more natural for users (including me) to `ssh` to the client and change the config file. AdminUI is great but `ssh` is better.
A reload on save would be helpful.

**Comment by fatedier** [comment]
I'm glad for your upcoming contribution.

Can we use `reload_on_update` instead of `reload_on_save` ?

The repo `https://github.com/fsnotify/fsnotify` maybe has some problem on macos like missing some events.

**Comment by YanB25** [comment]
> Can we use `reload_on_update` instead of `reload_on_save` ?

Okay.

> The repo `https://github.com/fsnotify/fsnotify` maybe has some problem on macos like missing some events.

The repo has some problems causing *extra* events on files ([this issue](https://github.com/howeyc/fsnotify/issues/62)). Users may find extra reload triggered in the log file. I couldn't find a stable library to watch a file on MacOS currently.

If extra reloads on MacOS is not acceptable, maybe we have to close this issue.

**Comment by fatedier** [comment]
Double events is ok. The problem is missing events.

You can try to test it or at least reload conf at a fixed interval ?

**Comment by YanB25** [comment]
> You can try to test it or at least reload conf at a fixed interval ?

Yes, I can get the `last modified time` at a fixed interval and reload on necessary, if you prefer to use polling.

> The problem is missing events.

I go through the issues and only find [this one](https://github.com/fsnotify/fsnotify/issues/220) related to losing events. There's no further information about it. Only double events bug is confirmed.

BTW, I have MacOS environment and I can have a test about it if necessary.

**Comment by fatedier** [comment]
@YanB25 No progress?

**Comment by zyfdegh** [comment]
Any update?

**Comment by fatedier** [comment]
This has been open for a long time — thanks for the suggestion. Since then, frpc has added a web dashboard with config editing, a reload API, and a store API for dynamic proxy/visitor management at runtime. These cover the original use case, so closing this.
